### PR TITLE
Add `substring` support for binary

### LINKED ISSUE #1593: Add `substring` support for `BinaryArray`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
We have supported `substring` for `StringArray`. And it is reasonable to add support for `BinaryArray`.
Arrow2 has implemented this.

This is one of tasks of #1531.

### PR DESCRIPTION
**Pull request by HaoYang670**
# Which issue does this PR close?

Closes #1593 .

# Rationale for this change

# What changes are included in this PR?
1. Add substring support for (Large)BinaryArray
2. Add some tests
3. fix a bug in the test of string array
4. update doc

# Are there any user-facing changes?
Update some docs

### CODE REVIEW

File: arrow/src/compute/kernels/substring.rs [hunk 1]
```diff
@@ -291,11 +575,14 @@ mod tests {
 
         cases.into_iter().try_for_each::<_, Result<()>>(
             |(array, start, length, expected)| {
-                let array = StringArray::from(array);
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/substring.rs:294]: We did not cover `LargeStringArray` in the past.

File: arrow/src/compute/kernels/substring.rs [hunk 2]
```diff
@@ -25,7 +26,68 @@ use crate::{
 };
 use std::cmp::Ordering;
 
-fn generic_substring<OffsetSize: StringOffsetSizeTrait>(
+fn binary_substring<OffsetSize: BinaryOffsetSizeTrait>(
```
Comment (HaoYang670) [inline arrow/src/compute/kernels/substring.rs:29]: The implementation is similar to `utf8_substring` expect that there is no char boundary checking.

Comment (alamb) [reply arrow/src/compute/kernels/substring.rs:29]: Rather than replicate quite so much code, I wonder if it would be possible to make `generic_substring` take a function to check char boundaries and and then pass in a function that does nothing

Maybe like(untested):

```rust
fn generic_substring<OffsetSize, F>(
    array: &GenericBinaryArray<OffsetSize>,
    start: OffsetSize,
    length: Option<OffsetSize>,
    check_char_boundary: F,
)
where 
  OffsetSize: StringOffsetSizeTrait,
  F: Fn(OffsetSize) -> Result<OffsetSize>
{
...
}

**Review by alamb — APPROVED** [review]
Thank you @HaoYang670  -- I went through the tests carefully and it looks good to me.

cc @viirya  and @jhorstmann

File: arrow/src/compute/kernels/substring.rs (hunk 2 above)
Comment (HaoYang670) [reply arrow/src/compute/kernels/substring.rs:29]: I am afraid that `BinaryArray` and `StringArray` can not share one API because `GenericBinaryArray` and `GenericStringArray` are two different types. 
Maybe we could use **macro** to extract some common codes.

Comment (alamb) [reply arrow/src/compute/kernels/substring.rs:29]: Got it -- makes sense -- duplication makes sense to me then

File: arrow/src/compute/kernels/substring.rs [hunk 3]
```diff
@@ -168,8 +246,214 @@ pub fn substring(array: &dyn Array, start: i64, length: Option<u64>) -> Result<A
 mod tests {
     use super::*;
 
-    fn with_nulls<T: 'static + Array + PartialEq + From<Vec<Option<&'static str>>>>(
-    ) -> Result<()> {
+    #[allow(clippy::type_complexity)]
+    fn with_nulls_generic_binary<O: BinaryOffsetSizeTrait>() -> Result<()> {
```
Comment (viirya) [inline arrow/src/compute/kernels/substring.rs:250]: I know this follows string version `with_nulls`, just wondering why in `with_nulls_...` only edge cases are tested against. `without_nulls_...` has normal test cases but it doesn't include nulls.

**Review by viirya — APPROVED** [review]
Looks good. Thanks @HaoYang670
